cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -92,45 +93,151 @@ func (c *machineController) findMachineSet(machine *v1alpha1.Machine) (*v1alpha1
 		return nil, fmt.Errorf("internal error; unexpected type: %T", machineSet)
 	}
 
+	if !machineIsOwnedByMachineSet(machine, machineSet) {
+		return nil, nil
+	}
+
 	return machineSet.DeepCopy(), nil
 }
 
-// run starts shared informers and waits for the shared informer cache
-// to synchronize.
+// run starts shared informers and waits for the informer cache to
+// synchronize.
 func (c *machineController) run(stopCh <-chan struct{}) error {
-	c.informerFactory.Start(stopCh)
+	c.kubeInformerFactory.Start(stopCh)
+	c.clusterInformerFactory.Start(stopCh)
+
+	glog.V(4).Infof("waiting for caches to sync")
+	if !cache.WaitForCacheSync(stopCh,
+		c.nodeInformer.HasSynced,
+		c.machineInformer.Informer().HasSynced,
+		c.machineSetInformer.Informer().HasSynced) {
+		return fmt.Errorf("syncing caches failed")
+	}
+
+	return nil
+}
 
-	glog.V(4).Infof("waiting for machine cache to sync")
-	if !cache.WaitForCacheSync(stopCh, c.machineInformer.Informer().HasSynced) {
-		return fmt.Errorf("cannot sync machine cache")
+// findMachineByNodeProviderID find associated machine using
+// node.Spec.ProviderID as the key. Returns nil if either the Node by
+// node.Spec.ProviderID cannot be found or if the node has no machine
+// annotation. A DeepCopy() of the object is returned on success.
+func (c *machineController) findMachineByNodeProviderID(node *apiv1.Node) (*v1alpha1.Machine, error) {
+	objs, err := c.nodeInformer.GetIndexer().ByIndex(nodeProviderIDIndex, node.Spec.ProviderID)
+	if err != nil {
+		return nil, err
 	}
 
-	glog.V(4).Infof("waiting for machineset cache to sync")
-	if !cache.WaitForCacheSync(stopCh, c.machineSetInformer.Informer().HasSynced) {
-		return fmt.Errorf("cannot sync machineset cache")
+	switch n := len(objs); {
+	case n == 0:
+		return nil, nil
+	case n > 1:
+		return nil, fmt.Errorf("internal error; expected len==1, got %v", n)
 	}
 
-	return nil
+	node, ok := objs[0].(*apiv1.Node)
+	if !ok {
+		return nil, fmt.Errorf("internal error; unexpected type %T", node)
+	}
+
+	machineName, found := node.Annotations["cluster.k8s.io/machine"]
+	if !found {
+		machineName, found = node.Annotations["machine"]
```

Comment:
q: we seem to fall back on this always currently, is there a separate pr that updates our annotation usage?